File: ai/pipeline/scoring.py
"""Lighting adequacy / risk scoring.

This is the single source of truth for the scoring model on the Python side.
The Go backend re-implements the identical formula in
`internal/domain/lighting/model/scoring.go` so that precomputed seed scores and
live-computed scores always agree.

Model (kept deliberately simple and explainable for the demo):

    density          = streetlights / (length_m / 100)      # fixtures per 100 m
    recommended      = RECOMMENDED_DENSITY[road_type]        # target fixtures / 100 m
    adequacy         = clamp(density / recommended, 0, 1)    # 0..1
    base_risk        = (1 - adequacy) * 100
    risk             = clamp(base_risk * road_weight * night_weight, 0, 100)

`road_weight` raises risk on busy roads (a dark highway is more dangerous than a
dark cul-de-sac). `night_weight` raises risk where most samples were captured at
night yet adequacy is still low (evidence the segment is actually dark in use).
"""

from __future__ import annotations

from dataclasses import dataclass

# Target streetlight fixtures per 100 m by road class (illumination engineering
# rules of thumb, simplified). Higher class road -> needs more / brighter light.
RECOMMENDED_DENSITY = {
    "highway": 4.0,
    "primary": 3.0,
    "secondary": 2.5,
    "residential": 1.8,
    "service": 1.2,
}

# How much an underlit road of each class contributes to risk.
ROAD_WEIGHT = {
    "highway": 1.25,
    "primary": 1.15,
    "secondary": 1.05,
    "residential": 0.95,
    "service": 0.85,
}

DEFAULT_ROAD_TYPE = "secondary"


def clamp(value: float, low: float, high: float) -> float:
    return max(low, min(high, value))


def lighting_density(streetlights: int, length_m: float) -> float:
    """Fixtures per 100 m of segment length."""
    if length_m <= 0:
        return 0.0
    return streetlights / (length_m / 100.0)


def recommended_density(road_type: str) -> float:
    return RECOMMENDED_DENSITY.get(road_type, RECOMMENDED_DENSITY[DEFAULT_ROAD_TYPE])


@dataclass
class RiskBreakdown:
    density: float
    recommended: float
    adequacy: float
    risk_score: float
    risk_level: str


def risk_level(score: float) -> str:
    if score >= 75:
        return "critical"
    if score >= 50:
        return "high"
    if score >= 25:
        return "medium"
    return "low"


def score_segment(
    streetlights: int,
    length_m: float,
    road_type: str,
    night_sample_ratio: float = 0.0,
) -> RiskBreakdown:
    density = lighting_density(streetlights, length_m)
    recommended = recommended_density(road_type)
    adequacy = clamp(density / recommended, 0.0, 1.0) if recommended > 0 else 1.0

    base_risk = (1.0 - adequacy) * 100.0
    road_weight = ROAD_WEIGHT.get(road_type, ROAD_WEIGHT[DEFAULT_ROAD_TYPE])
    # Night weight: up to +20% risk when a segment is mostly observed at night
    # but still inadequately lit.
    night_weight = 1.0 + 0.20 * clamp(night_sample_ratio, 0.0, 1.0) * (1.0 - adequacy)

    risk = clamp(base_risk * road_weight * night_weight, 0.0, 100.0)
    return RiskBreakdown(
        density=round(density, 3),
        recommended=recommended,
        adequacy=round(adequacy, 3),
        risk_score=round(risk, 1),
        risk_level=risk_level(risk),
    )

File: backend/internal/domain/lighting/model/scoring.go
package model

// Lighting adequacy / risk scoring. This is a 1:1 port of
// ai/pipeline/scoring.py so seed scores (computed in Python) and live scores
// (computed here in Go) always agree.
//
//	density     = streetlights / (length_m / 100)
//	recommended = RecommendedDensity[road_type]
//	adequacy    = clamp(density / recommended, 0, 1)
//	base_risk   = (1 - adequacy) * 100
//	risk        = clamp(base_risk * road_weight * night_weight, 0, 100)

// RecommendedDensity is the target streetlight fixtures per 100 m by road class.
var RecommendedDensity = map[string]float64{
	"highway":     4.0,
	"primary":     3.0,
	"secondary":   2.5,
	"residential": 1.8,
	"service":     1.2,
}

// RoadWeight scales how much an underlit road of each class contributes to risk.
var RoadWeight = map[string]float64{
	"highway":     1.25,
	"primary":     1.15,
	"secondary":   1.05,
	"residential": 0.95,
	"service":     0.85,
}

// DefaultRoadType is used when a segment's road type is unknown.
const DefaultRoadType = "secondary"

// RiskBreakdown is the full scoring output for a segment or point.
type RiskBreakdown struct {
	Density     float64 `json:"lighting_density"`
	Recommended float64 `json:"recommended_density"`
	Adequacy    float64 `json:"adequacy"`
	RiskScore   float64 `json:"risk_score"`
	RiskLevel   string  `json:"risk_level"`
}

func clamp(v, lo, hi float64) float64 {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

func round1(v float64) float64 {
	return float64(int(v*10+0.5)) / 10
}

func round3(v float64) float64 {
	return float64(int(v*1000+0.5)) / 1000
}

// LightingDensity returns fixtures per 100 m of segment length.
func LightingDensity(streetlights int, lengthM float64) float64 {
	if lengthM <= 0 {
		return 0
	}
	return float64(streetlights) / (lengthM / 100.0)
}

// RecommendedDensityFor returns the target density for a road type.
func RecommendedDensityFor(roadType string) float64 {
	if v, ok := RecommendedDensity[roadType]; ok {
		return v
	}
	return RecommendedDensity[DefaultRoadType]
}

// RoadWeightFor returns the risk weight for a road type.
func RoadWeightFor(roadType string) float64 {
	if v, ok := RoadWeight[roadType]; ok {
		return v
	}
	return RoadWeight[DefaultRoadType]
}

// RiskLevelFor buckets a numeric risk score.
func RiskLevelFor(score float64) string {
	switch {
	case score >= 75:
		return string(RiskCritical)
	case score >= 50:
		return string(RiskHigh)
	case score >= 25:
		return string(RiskMedium)
	default:
		return string(RiskLow)
	}
}

// ScoreSegment computes the full risk breakdown for a segment.
func ScoreSegment(streetlights int, lengthM float64, roadType string, nightRatio float64) RiskBreakdown {
	density := LightingDensity(streetlights, lengthM)
	recommended := RecommendedDensityFor(roadType)
	adequacy := 1.0
	if recommended > 0 {
		adequacy = clamp(density/recommended, 0, 1)
	}
	baseRisk := (1.0 - adequacy) * 100.0
	roadWeight := RoadWeightFor(roadType)
	nightWeight := 1.0 + 0.20*clamp(nightRatio, 0, 1)*(1.0-adequacy)
	risk := clamp(baseRisk*roadWeight*nightWeight, 0, 100)

	return RiskBreakdown{
		Density:     round3(density),
		Recommended: recommended,
		Adequacy:    round3(adequacy),
		RiskScore:   round1(risk),
		RiskLevel:   RiskLevelFor(risk),
	}
}
